OIDC auth: MPCONFIG provisioning, PKCE support and integration tests (#9273)
IQSS/dataverse: doc/sphinx-guides/source/installation/oidc.rst

.. _oidc-pkce:

Enabling PKCE Security
^^^^^^^^^^^^^^^^^^^^^^

Many providers these days support or even require the usage of `PKCE <https://oauth.net/2/pkce/>`_ to safeguard against
some attacks and enable public clients that cannot have a secure secret to still use OpenID Connect (or OAuth2).

The Dataverse-built OIDC client can be configured to use PKCE and the method to use when creating the code challenge can be specified.
See also `this explanation of the flow <https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow-with-proof-key-for-code-exchange-pkce>`_
for details on how this works.

As we are using the `Nimbus SDK <https://connect2id.com/products/nimbus-oauth-openid-connect-sdk>`_ as our client
library, we support the standard ``PLAIN`` and ``S256`` (SHA-256) code challenge methods. "SHA-256 method" is the default
as recommend in `RFC7636 <https://datatracker.ietf.org/doc/html/rfc7636#section-4.2>`_. If your provider needs some
other method, please open an issue.

The provisioning sections below contain in the example the parameters you may use to configure PKCE.

> Provision a provider

Reviewer: ```suggestion
Provision a Provider
```

Also can you please change these to Title Case?

- How to use
- Other use cases and combinations

Author: Done in 274f0e98

some attacks and enable public clients that cannot have a secure secret to still use OpenID Connect (or OAuth2).

The Dataverse built OIDC client can be enabled to use PKCE and which method to use when creating the code challenge.
See also `this explanation of the flow <https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow-with-proof-key-for-code-exchange-pkce>`_
for details on how this works.

As we are using the `Nimbus SDK <https://connect2id.com/products/nimbus-oauth-openid-connect-sdk>`_ as our client
library, we support the standard ``PLAIN`` and ``S256`` code challenge methods. "SHA-256 method" is the default
as recommend in `RFC7636 <https://datatracker.ietf.org/doc/html/rfc7636#section-4.2>`_. If your provider needs some
other method (unlikely), please open an issue.

The provisioning sections below contain in the example the parameters you may use to configure PKCE.

Provision via REST API
^^^^^^^^^^^^^^^^^^^^^^

Note: you may omit the PKCE related settings from ``factoryData`` below if you don't plan on using PKCE - default is
disabled.

> Please create a ``my-oidc-provider.json`` file like this, replacing every ``<...>`` with your values:

Author: Done in [bb0ba9a](https://github.com/IQSS/dataverse/pull/9273/commits/bb0ba9abfd934415b6a8a2c24da3cab1f279b291). Resolving here, please review the new text. Thx!

You should see the new provider under "Other options" on the Log In page, as described in the :doc:`/user/account`
section of the User Guide.

> Provision via MPCONFIG

Reviewer: Can we add a link from `installation/config.rst` to this list? Otherwise I fear people will complain that there's no single list of MPCONFIG options.

On a related notes, it's hard to find all the docs for APIs so for now at least we list where to find all the scattered pages here: https://guides.dataverse.org/en/5.13/api/intro.html#lists-of-dataverse-apis

Author: Done in [edef7d5](https://github.com/IQSS/dataverse/pull/9273/commits/edef7d5bd52f3625b2e15521046d97a73f999894). See https://dataverse-guide--9273.org.readthedocs.build/en/9273/installation/config.html#dataverse-auth-oidc

You should see the new provider under "Other options" on the Log In page, as described in the :doc:`/user/account`
section of the User Guide.

.. _oidc-mpconfig:

Provision via MPCONFIG
^^^^^^^^^^^^^^^^^^^^^^

> In case you only require a single OIDC provider, you can also provision an OIDC based provider using MicroProfile Config

Reviewer: Is it premature to consider supporting multiple OIDC providers, perhaps like how we support multiple file stores?

Reviewer 2: I agree with Oliver's comment, especially if the authorization (groups, ACLs) is also managed by an "OIDC server".

Author: @pdurbin can we resolve this?

sources. Note that this provider will only be deployed at startup time and (currently) cannot be reconfigured without
a restart.

The following options are available:

.. list-table::
  :widths: 25 55 10 10
  :header-rows: 1
  :align: left

  * - Option
    - Description
    - Mandatory
    - Default
  * - ``dataverse.auth.oidc.enabled``
    - Enable or disable provisioning the provider via MicroProfile.
    - N
    - ``false``
  * - ``dataverse.auth.oidc.client-id``
    - The client-id of the application to identify it at your provider.
    - Y
    - \-
  * - ``dataverse.auth.oidc.client-secret``
    - A confidential secret to authorize application requests to the provider as legit.
    - N
    - \-
  * - ``dataverse.auth.oidc.auth-server-url``
    - The base URL of the OpenID Connect (OIDC) server as explained above.
    - Y
    - \-
  * - ``dataverse.auth.oidc.pkce.enabled``
    - Set to ``true`` to enable :ref:`PKCE <oidc-pkce>` in auth flow.
    - N
    - ``false``
  * - ``dataverse.auth.oidc.pkce.method``

>     - Set code challenge method. Default equals best practice.

Author: @pdurbin I'm not sure I agree with the change. What I was trying to say seems to be different than what your suggestions expresses. I was trying to express in a very condensed way that the default value in place will be the value which is best practice in the literature (and thus the one usually in place, which is why it's the default)

>     - Set code challenge method. Default equals best practice.

Reviewer: Ok. Maybe more words will help. I added another ```suggestion.

You should see the new provider under "Other options" on the Log In page, as described in the :doc:`/user/account`
section of the User Guide.

.. _oidc-mpconfig:

Provision via MPCONFIG
^^^^^^^^^^^^^^^^^^^^^^

> In case you only require a single OIDC provider, you can also provision an OIDC based provider using MicroProfile Config

Reviewer: This looks great but I'm a little concerned that a new user will be confused about MPCONFIG and how to configure it. Either in this PR or a future one, maybe we can link to a blurb about this (on the main config.rst page, I assume).

Author: Done with a2b9d56f

sources. Note that this provider will only be deployed at startup time and (currently) cannot be reconfigured without
a restart.

The following options are available:

.. list-table::
  :widths: 25 55 10 10
  :header-rows: 1
  :align: left

  * - Option
    - Description
    - Mandatory
    - Default
  * - ``dataverse.auth.oidc.enabled``
    - Enable or disable provisioning the provider via MicroProfile.
    - N
    - ``false``
  * - ``dataverse.auth.oidc.client-id``
    - The client-id of the application to identify it at your provider.
    - Y
    - \-
  * - ``dataverse.auth.oidc.client-secret``
    - A confidential secret to authorize application requests to the provider as legit.
    - N
    - \-
  * - ``dataverse.auth.oidc.auth-server-url``
    - The base URL of the OpenID Connect (OIDC) server as explained above.
    - Y
    - \-
  * - ``dataverse.auth.oidc.pkce.enabled``
    - Set to ``true`` to enable :ref:`PKCE <oidc-pkce>` in auth flow.
    - N
    - ``false``
  * - ``dataverse.auth.oidc.pkce.method``

>     - Set code challenge method. Default equals best practice.

Reviewer: ```suggestion
    - Set code challenge method. The default value is the current best practice in the literature.
```
